sensing/pointcloud_preprocessor/src/outlier_filter/ring_outlier_filter_nodelet.cpp

```diff
@@ -44,82 +44,116 @@ RingOutlierFilterComponent::RingOutlierFilterComponent(const rclcpp::NodeOptions
+// TODO(sykwer): Temporary Implementation: Rename this function to `filter()` when all the filter
+// nodes conform to new API. Then delete the old `filter()` defined below.
+void RingOutlierFilterComponent::faster_filter(
   const PointCloud2ConstPtr & input, [[maybe_unused]] const IndicesPtr & indices,
-  PointCloud2 & output)
+  PointCloud2 & output, const TransformInfo & transform_info)
 {
+  (void)transform_info;
+
   std::scoped_lock lock(mutex_);
   stop_watch_ptr_->toc("processing_time", true);
-  std::unordered_map<uint16_t, std::vector<std::size_t>> input_ring_map;
-  input_ring_map.reserve(128);
-  sensor_msgs::msg::PointCloud2::SharedPtr input_ptr =
-    std::make_shared<sensor_msgs::msg::PointCloud2>(*input);
+
+  output.point_step = sizeof(PointXYZI);
+  output.data.resize(output.point_step * input->width);
+  size_t output_size = 0;
 
   const auto ring_offset =
     input->fields.at(static_cast<size_t>(autoware_point_types::PointIndex::Ring)).offset;
-  for (std::size_t idx = 0U; idx < input_ptr->data.size(); idx += input_ptr->point_step) {
-    input_ring_map[*reinterpret_cast<uint16_t *>(&input_ptr->data[idx + ring_offset])].push_back(
-      idx);
-  }
-
-  PointCloud2Modifier<PointXYZI> output_modifier{output, input->header.frame_id};
-  output_modifier.reserve(input->width);
-
-  std::vector<std::size_t> tmp_indices;
-  tmp_indices.reserve(input->width);
-
   const auto azimuth_offset =
     input->fields.at(static_cast<size_t>(autoware_point_types::PointIndex::Azimuth)).offset;
   const auto distance_offset =
     input->fields.at(static_cast<size_t>(autoware_point_types::PointIndex::Distance)).offset;
-  for (const auto & ring_indices : input_ring_map) {
-    if (ring_indices.second.size() < 2) {
-      continue;
-    }
 
-    for (size_t idx = 0U; idx < ring_indices.second.size() - 1; ++idx) {
-      const auto & current_idx = ring_indices.second.at(idx);
-      const auto & next_idx = ring_indices.second.at(idx + 1);
-      tmp_indices.emplace_back(current_idx);
+  std::vector<std::vector<size_t>> ring2indices;
+  ring2indices.reserve(128);
+
+  for (size_t i = 0; i < 128; i++) {
+    ring2indices.push_back(std::vector<size_t>());
+    ring2indices.back().reserve(2000);  // TODO(sykwer): Make it parameter
+  }
+
+  for (size_t data_idx = 0; data_idx < input->data.size(); data_idx += input->point_step) {
+    const uint16_t ring = *reinterpret_cast<const uint16_t *>(&input->data[data_idx + ring_offset]);
+    ring2indices[ring].push_back(data_idx);
+  }
+
+  // walk range: [walk_first_idx, walk_last_idx]
+  int walk_first_idx = 0;
+  int walk_last_idx = -1;
+
+  for (const auto & indices : ring2indices) {
+    if (indices.size() < 2) continue;
+
+    walk_first_idx = 0;
+    walk_last_idx = -1;
+
+    for (size_t idx = 0U; idx < indices.size() - 1; ++idx) {
+      const size_t & current_data_idx = indices[idx];
+      const size_t & next_data_idx = indices[idx + 1];
+      walk_last_idx = idx;
```

Comment:
I think we are skipping the last point with this logic.
For example, let's imagine a perfect circle ring with N points.

When we first enter the first loop we have :
```
idx = 0
walk_first_idx = 0
walk_last_idx = idx = 0
current_data_idx = indices[0]
next_data_idx= indices[1]
```

The ring is a perfect circle, so all the points are part of the same walk. So we always continue here:
```c++
      if (
        std::max(current_distance, next_distance) <
          std::min(current_distance, next_distance) * distance_ratio_ &&
        azimuth_diff < 100.f) {
        continue;  // Determined to be included in the same walk
      }
```

On the last loop we have:
```
idx = N-2
walk_first_idx = 0
walk_last_idx = idx = N-2
current_data_idx = indices[N-2]
next_data_idx= indices[N-1]
```

So after the loop, we have these lines:
```c++
  if (isCluster(
          input, indices[walk_first_idx], indices[walk_last_idx],
          walk_last_idx - walk_first_idx + 1)) {
      for (int i = walk_first_idx; i <= walk_last_idx; i++) {
        auto output_ptr = reinterpret_cast<PointXYZI *>(&output.data[output_size]);
        *output_ptr = *reinterpret_cast<const PointXYZI *>(&input->data[indices[i]]);
        output_size += output.point_step;
      }
    }
  }
```
which basically copy points from `walk_first_idx = 0` to `walk_last_idx = N-2`, thus skipping the last point.

Reply:
same as below